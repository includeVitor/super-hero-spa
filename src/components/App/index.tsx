import { BrowserRouter } from 'react-router-dom'
import AppRouter from '../../route'
import { Main } from './styles'

const App = () => {
  return (
    <BrowserRouter>
      <Main>
        <AppRouter />
      </Main>
    </BrowserRouter>
  )
}

export default App
