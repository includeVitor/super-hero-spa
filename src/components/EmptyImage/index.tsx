import { memo } from 'react'
import { Article } from './styles'

const EmptyImage = () => <Article>Image could not be loaded</Article>

export default memo(EmptyImage)
