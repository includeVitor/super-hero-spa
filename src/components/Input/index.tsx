import { InputHTMLAttributes, memo } from 'react'
import { Container, InputStyled } from './styles'

type InputProps = {} & InputHTMLAttributes<HTMLInputElement>

const Input = ({ ...rest }: InputProps) => {
  return (
    <Container>
      <InputStyled autoComplete="off" spellCheck="false" {...rest} />
    </Container>
  )
}

export default memo(Input)
