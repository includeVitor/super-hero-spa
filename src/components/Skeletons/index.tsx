import Skeleton from 'react-loading-skeleton'
import { LoadingCharactersArticle } from './styles'

const LoadingCharacters = () => {
  return (
    <LoadingCharactersArticle>
      <Skeleton height="22rem" width="15rem" />
      <Skeleton height="22rem" width="15rem" />
      <Skeleton height="22rem" width="15rem" />
      <Skeleton height="22rem" width="15rem" />
      <Skeleton height="22rem" width="15rem" />
      <Skeleton height="22rem" width="15rem" />
      <Skeleton height="22rem" width="15rem" />
      <Skeleton height="22rem" width="15rem" />
    </LoadingCharactersArticle>
  )
}

export { LoadingCharacters }
