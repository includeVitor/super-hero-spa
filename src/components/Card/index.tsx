import { memo, useCallback, useState } from 'react'
import EmptyImage from '../EmptyImage'
import { Article, ButtonAlign, Description, Image, Title } from './styles'

type CardProps = {
  title: string
  description?: string
  url: string
  $selected?: boolean
  handleView?: () => void
  handleCombat?: () => void
}

const Card = ({
  title,
  description,
  url,
  $selected,
  handleView,
  handleCombat
}: CardProps) => {
  const [imgLoaded, setImgLoaded] = useState(true)

  const handleImgLoadError = useCallback(() => {
    setImgLoaded(false)
  }, [])

  return (
    <Article $selected={$selected}>
      <Title>{title}</Title>
      <Description>{description}</Description>
      {imgLoaded ? (
        <Image src={url} onError={handleImgLoadError} />
      ) : (
        <EmptyImage />
      )}

      <ButtonAlign>
        {handleView && <button onClick={handleView}>View</button>}
        {handleCombat && <button onClick={handleCombat}>Combat</button>}
      </ButtonAlign>
    </Article>
  )
}

export default memo(Card)
