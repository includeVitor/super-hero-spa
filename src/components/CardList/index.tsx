import { memo, useCallback } from 'react'
import { useNavigate } from 'react-router-dom'
import { sumPowers } from '../../lib/utils'
import { AppRoutes } from '../../route/types'
import { Character } from '../../service/superHero/types'
import Card from '../Card'
import { PowerStats, Section } from './styles'

type CardlistProps = {
  characters: Character[]
  $powerstats?: boolean
  setCharacters: React.Dispatch<React.SetStateAction<Character[]>>
  setSelectedCharacters: React.Dispatch<React.SetStateAction<Character[]>>
}

const CardList = ({
  characters,
  $powerstats = false,
  setCharacters,
  setSelectedCharacters
}: CardlistProps) => {
  const navigate = useNavigate()

  const handleCombat = useCallback((id: string) => {
    setCharacters(prev =>
      prev.map(p => (p.id === id ? { ...p, $selected: !p.$selected } : p))
    )
    setSelectedCharacters(prev => [...prev, characters.find(f => f.id === id)!])
  }, [])

  const handleView = useCallback((id: string) => {
    navigate(AppRoutes.Details, {
      replace: true,
      state: { character: characters.find(f => f.id === id)! }
    })
  }, [])

  const list = characters.map(character => (
    <PowerStats $powerstats={$powerstats} key={character.id}>
      {$powerstats && (
        <ul>
          <li>Total: {sumPowers(character.powerstats)}</li>
          <li>Intelligence: {character.powerstats.intelligence}</li>
          <li>Strength: {character.powerstats.strength}</li>
          <li>Speed: {character.powerstats.speed}</li>
          <li>Durability: {character.powerstats.durability}</li>
          <li>Power: {character.powerstats.power}</li>
          <li>Combat: {character.powerstats.combat}</li>
        </ul>
      )}
      <Card
        key={character.id}
        id={character.id}
        title={character.name}
        description={character.biography['full-name']}
        url={character.image.url}
        handleCombat={handleCombat}
        handleView={handleView}
        $selected={character.$selected}
      />
    </PowerStats>
  ))

  return <Section>{list}</Section>
}

export default memo(CardList)
