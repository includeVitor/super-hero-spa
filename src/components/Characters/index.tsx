import { memo } from 'react'
import { Character } from '../../service/superHero/types'
import Card from '../Card'
import { Section } from './styles'

type CharactersProps = {
  characters: Pick<Character, 'id' | 'name' | 'image' | 'biography'>[]
}

const Characters = ({ characters }: CharactersProps) => {
  const list = characters.map(character => (
    <Card
      key={character.id}
      title={character.name}
      description={character.biography['full-name']}
      url={character.image.url}
    />
  ))

  return <Section>{list}</Section>
}

export default memo(Characters)
