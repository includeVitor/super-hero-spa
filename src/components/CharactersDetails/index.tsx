import { useNavigate } from 'react-router-dom'
import { AppRoutes } from '../../route/types'
import { Character } from '../../service/superHero/types'
import Card from '../Card'
import { Section } from './styles'

type CharactersDetailsProps = {
  character: Character
}

const CharactersDetails = ({ character }: CharactersDetailsProps) => {
  const navigate = useNavigate()

  const handleClick = () => navigate(AppRoutes.HomePage)

  return (
    <>
      <button onClick={handleClick}>go back</button>
      <Section>
        <Card
          key={character.id}
          title={character.name}
          description={character.biography['full-name']}
          url={character.image.url}
        />

        <div style={{ width: '500px', height: '500px' }}>
          <div>Id: {character.id}</div>
          <div>Name: {character.name}</div>

          <h3>PowerStats</h3>
          <ul>
            <li>Intelligence: {character.powerstats.intelligence}</li>
            <li>Strength: {character.powerstats.strength}</li>
            <li>Speed: {character.powerstats.speed}</li>
            <li>Durability: {character.powerstats.durability}</li>
            <li>Power: {character.powerstats.power}</li>
            <li>Combat: {character.powerstats.combat}</li>
          </ul>

          <h3>Biography</h3>
          <ul>
            <li>Nome Completo: {character.biography['full-name']}</li>
            <li>Alter Egos: {character.biography['alter-egos']}</li>
            <li>Place of Birth: {character.biography['place-of-birth']}</li>
            <li>First Appearance: {character.biography['first-appearance']}</li>
            <li>Publisher: {character.biography.publisher}</li>
            <li>Alignment: {character.biography.alignment}</li>
          </ul>

          <h3>Appearance</h3>
          <ul>
            <li>Gender: {character.appearance.gender}</li>
            <li>Race: {character.appearance.race}</li>
            <li>Height: {character.appearance.height}</li>
            <li>Weight: {character.appearance.weight}</li>
            <li>Eye Color: {character.appearance['eye-color']}</li>
            <li>Hair Color: {character.appearance['hair-color']}</li>
          </ul>

          <h3>Work</h3>
          <ul>
            <li>Occupation: {character.work.occupation}</li>
            <li>Base: {character.work.base}</li>
          </ul>

          <h3>Connections</h3>
          <ul>
            <li>
              GroupAffiliation: {character.connections['group-affiliation']}
            </li>
            <li>Relatives: {character.connections.relatives}</li>
          </ul>
        </div>
      </Section>
    </>
  )
}

export default CharactersDetails
