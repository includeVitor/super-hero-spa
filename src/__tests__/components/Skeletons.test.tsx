import { render } from '@testing-library/react'
import { LoadingCharacters } from '../../components/Skeletons'

describe('Skeletons', () => {
  it('LoadingCharacters should match the snapshot', () => {
    const { container } = render(<LoadingCharacters />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
