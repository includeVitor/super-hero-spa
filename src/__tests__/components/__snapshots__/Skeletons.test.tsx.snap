// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeletons LoadingCharacters should match the snapshot 1`] = `
<article
  class="sc-bdvvtL kPhNLK"
>
  <span
    aria-busy="true"
    aria-live="polite"
  >
    <span
      class="react-loading-skeleton"
      style="width: 15rem; height: 22rem;"
    >
      ‌
    </span>
    <br />
  </span>
  <span
    aria-busy="true"
    aria-live="polite"
  >
    <span
      class="react-loading-skeleton"
      style="width: 15rem; height: 22rem;"
    >
      ‌
    </span>
    <br />
  </span>
  <span
    aria-busy="true"
    aria-live="polite"
  >
    <span
      class="react-loading-skeleton"
      style="width: 15rem; height: 22rem;"
    >
      ‌
    </span>
    <br />
  </span>
  <span
    aria-busy="true"
    aria-live="polite"
  >
    <span
      class="react-loading-skeleton"
      style="width: 15rem; height: 22rem;"
    >
      ‌
    </span>
    <br />
  </span>
  <span
    aria-busy="true"
    aria-live="polite"
  >
    <span
      class="react-loading-skeleton"
      style="width: 15rem; height: 22rem;"
    >
      ‌
    </span>
    <br />
  </span>
  <span
    aria-busy="true"
    aria-live="polite"
  >
    <span
      class="react-loading-skeleton"
      style="width: 15rem; height: 22rem;"
    >
      ‌
    </span>
    <br />
  </span>
  <span
    aria-busy="true"
    aria-live="polite"
  >
    <span
      class="react-loading-skeleton"
      style="width: 15rem; height: 22rem;"
    >
      ‌
    </span>
    <br />
  </span>
  <span
    aria-busy="true"
    aria-live="polite"
  >
    <span
      class="react-loading-skeleton"
      style="width: 15rem; height: 22rem;"
    >
      ‌
    </span>
    <br />
  </span>
</article>
`;
